silkworm/db/snapshots/index/bloom_filter.cpp

```diff
@@ -0,0 +1,206 @@
+/*
+   Copyright 2024 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include "bloom_filter.hpp"
+
+#include <numbers>
+#include <random>
+#include <sstream>
+#include <string_view>
+
+#include <openssl/evp.h>
+
+#include <silkworm/core/common/bytes_to_string.hpp>
+#include <silkworm/core/common/endian.hpp>
+#include <silkworm/core/common/util.hpp>
+
+namespace silkworm::snapshots::index {
+
+using namespace std::numbers;
+
+//! The minimum Bloom filter bits count
+static constexpr size_t kMinimumBitsCount{2};
+
+//! kRotation sets how much to rotate the hash on each filter iteration.
+//! This is somewhat randomly set to a prime on the lower segment of 64.
+constexpr size_t kRotation{17};
+
+constexpr size_t kRotationOf64{64 - kRotation};
+
+//! The magic header used in serialization format for version v2
+static const ByteView kMagicHeader{string_view_to_byte_view("\0\0\0\0\0\0\0\0v02\n"sv)};
+
+uint64_t BloomFilter::optimal_bits_count(uint64_t max_key_count, double p) {
+    return static_cast<uint64_t>(std::ceil(-static_cast<double>(max_key_count) * std::log(p) / (ln2 * ln2)));
+}
+
+std::unique_ptr<BloomFilter> BloomFilter::read_from(std::istream& index_input_stream) {
+    auto filter = std::make_unique<BloomFilter>(kMinimumBitsCount);  // bits count doesn't matter here, it will be overwritten
+    index_input_stream >> *filter;
+    return filter;
+}
```

Comment:
Given that you've implemented a read operator, I suggest to remove this wrapper method, and use it the operator instead. kMinimumBitsCount could be a default value to be able to default construct a BloomFilter object.

I'm also not against the inverse: remove `>>` operator, and put all logic into read_from.
Having both seems redundant.

Reply:
I'll remove `BloomFilter::read_from`